### PR TITLE
KBV-93 - KBV CRI API /question resource

### PR DESCRIPTION
**Pull request by amanm-alphagov**
### What changed

Implemented the `/question` resource for the KBV API. This API is expected to be called by the KBV frontend. 

### Issue tracking

- [KBV-93](https://govukverify.atlassian.net/browse/KBV-93)

### Environment variables or secrets

- [x] No environment variables or secrets were added or changed

- [ ] Documented in the [README](./blob/main/README.md)
- [ ] Added to deployment repository
- [ ] Added to local startup repository

### Other considerations

- [ ] Update [README](./blob/main/README.md) with any new instructions or tasks

### CODE REVIEW

File: lambdas/session/src/main/java/uk/gov/di/ipv/cri/kbv/api/handler/QuestionAnswerHandler.java [hunk 1]
```diff
@@ -8,10 +8,7 @@
 import com.fasterxml.jackson.databind.ObjectMapper;
 import software.amazon.lambda.powertools.tracing.CaptureMode;
 import software.amazon.lambda.powertools.tracing.Tracing;
-import uk.gov.di.ipv.cri.kbv.api.domain.QuestionAnswer;
-import uk.gov.di.ipv.cri.kbv.api.domain.QuestionAnswerRequest;
-import uk.gov.di.ipv.cri.kbv.api.domain.QuestionState;
-import uk.gov.di.ipv.cri.kbv.api.domain.QuestionsResponse;
+import uk.gov.di.ipv.cri.kbv.api.domain.*;
```
Comment (jkunle) [inline lambdas/session/src/main/java/uk/gov/di/ipv/cri/kbv/api/handler/QuestionAnswerHandler.java:11]: * you used star import here see https://gds-way.cloudapps.digital/manuals/programming-languages/java.html#imports

File: lambdas/session/src/main/java/uk/gov/di/ipv/cri/kbv/api/handler/QuestionAnswerHandler.java [hunk 2]
```diff
@@ -45,7 +42,7 @@ public APIGatewayProxyResponseEvent handleRequest(
         String responseBody = "{}";
         Map<String, String> responseHeaders = Map.of("Content-Type", "application/json");
         String sessionId = input.getHeaders().get(HEADER_SESSION_ID);
-        KBVSessionItem kbvSessionItem = storageService.getSessionId(sessionId);
+        KBVSessionItem kbvSessionItem = storageService.getSessionId(sessionId).orElseThrow();
         QuestionState questionState = null;
         try {
```
Comment (jkunle) [inline lambdas/session/src/main/java/uk/gov/di/ipv/cri/kbv/api/handler/QuestionAnswerHandler.java:47]: There seems to be too many try blocks here maybe just one would be ok. It have multi catch blocks if needed

File: lambdas/question/src/main/java/uk.gov.di.ipv.cri.kbv.api.handler/QuestionHandler.java [hunk 3]
```diff
@@ -43,83 +46,83 @@ public QuestionHandler() {
...
+                            questionState.getControl().getURN());
                     nextQuestion = questionState.getNextQuestion();
                     responseBody = objectMapper.writeValueAsString(nextQuestion.get());
                 }
             } else {
                 // TODO Handle scenario when no questions are available
-                statusCode = 200;
+                statusCode = HttpStatus.SC_OK;
+                nextQuestion = questionState.getNextQuestion();
+                if (!nextQuestion.isEmpty()) {
```
Comment (jkunle) [inline lambdas/question/src/main/java/uk.gov.di.ipv.cri.kbv.api.handler/QuestionHandler.java:104]: nextQuestion.isPresent

File: lambdas/question/src/test/java/uk.gov.di.ipv.cri.kbv.api.handler/QuestionHandlerTest.java [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+import org.slf4j.LoggerFactory;
+import uk.gov.di.ipv.cri.kbv.api.domain.*;
+import uk.gov.di.ipv.cri.kbv.api.persistence.item.KBVSessionItem;
```
Comment (jkunle) [inline lambdas/question/src/test/java/uk.gov.di.ipv.cri.kbv.api.handler/QuestionHandlerTest.java:23]: * see https://gds-way.cloudapps.digital/manuals/programming-languages/java.html#imports

File: lambdas/question/src/test/java/uk.gov.di.ipv.cri.kbv.api.handler/QuestionHandlerTest.java [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+import uk.gov.di.ipv.cri.kbv.api.service.ExperianService;
+import uk.gov.di.ipv.cri.kbv.api.service.StorageService;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Optional;
+
+import static org.junit.jupiter.api.Assertions.*;
```
Comment (jkunle) [inline lambdas/question/src/test/java/uk.gov.di.ipv.cri.kbv.api.handler/QuestionHandlerTest.java:31]: single line imports to replace * imports see above

File: lambdas/question/src/test/java/uk.gov.di.ipv.cri.kbv.api.handler/QuestionHandlerTest.java [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+        when(mockObjectMapper.writeValueAsString(personIdentityMock)).thenReturn("person-identity");
+
+        QuestionsResponse questionsResponseMock = mock(QuestionsResponse.class);
+        when(mockExperianService.getQuestions("person-identity")).thenReturn(questionsResponseMock);
+        when(questionStateMock.setQuestionsResponse(questionsResponseMock)).thenReturn(true);
+        String state = "question-state";
+        when(mockObjectMapper.writeValueAsString(questionStateMock)).thenReturn(state);
+        Control controlMock = mock(Control.class);
+        when(questionStateMock.getControl()).thenReturn(controlMock);
+        when(controlMock.getAuthRefNo()).thenReturn("auth-ref-no");
```
Comment (jkunle) [inline lambdas/question/src/test/java/uk.gov.di.ipv.cri.kbv.api.handler/QuestionHandlerTest.java:87]: use variables here i.e `String authRefNo = "auth-ref-no"`

File: lambdas/question/src/test/java/uk.gov.di.ipv.cri.kbv.api.handler/QuestionHandlerTest.java [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+        when(controlMock.getURN()).thenReturn("ipv-session-id");
+        doNothing()
+                .when(mockStorageService)
+                .update(
+                        sessionHeader.get(HEADER_SESSION_ID),
+                        state,
+                        "auth-ref-no",
+                        "ipv-session-id");
```
Comment (jkunle) [inline lambdas/question/src/test/java/uk.gov.di.ipv.cri.kbv.api.handler/QuestionHandlerTest.java:95]: same here

File: lambdas/session/src/main/java/uk/gov/di/ipv/cri/kbv/api/handler/QuestionAnswerHandler.java (hunk 1 above)
Comment (jkunle) [inline lambdas/session/src/main/java/uk/gov/di/ipv/cri/kbv/api/handler/QuestionAnswerHandler.java:11]: use single line import here

File: lambdas/question/src/test/java/uk.gov.di.ipv.cri.kbv.api.handler/QuestionHandlerTest.java (hunk 7 above)
Comment (jkunle) [reply lambdas/question/src/test/java/uk.gov.di.ipv.cri.kbv.api.handler/QuestionHandlerTest.java:95]: Try and address some of the 20 code smells highlighted by SonarCloud

File: lambdas/question/src/main/java/uk.gov.di.ipv.cri.kbv.api.handler/QuestionHandler.java (hunk 3 above)
Comment (amanm-alphagov) [reply lambdas/question/src/main/java/uk.gov.di.ipv.cri.kbv.api.handler/QuestionHandler.java:104]: Changed handler and handler test code to implement the same

File: lambdas/question/src/test/java/uk.gov.di.ipv.cri.kbv.api.handler/QuestionHandlerTest.java (hunk 5 above)
Comment (amanm-alphagov) [reply lambdas/question/src/test/java/uk.gov.di.ipv.cri.kbv.api.handler/QuestionHandlerTest.java:31]: Done

File: lambdas/question/src/test/java/uk.gov.di.ipv.cri.kbv.api.handler/QuestionHandlerTest.java (hunk 6 above)
Comment (amanm-alphagov) [reply lambdas/question/src/test/java/uk.gov.di.ipv.cri.kbv.api.handler/QuestionHandlerTest.java:87]: Done

File: template.yaml [hunk 8]
```diff
@@ -127,21 +140,67 @@ Resources:
...
+      MemorySize: 512
+      Tracing: Active
+      Environment:
+        Variables:
+          IS_LOCAL: false
+          KBV_SESSION_TABLE_NAME: !Ref KBVSessionTable
+          POWERTOOLS_SERVICE_NAME: di-ipv-cri-kbv-api
+          EXPERIAN_API_WRAPPER_SAA_RESOURCE: /question-request
+          EXPERIAN_API_WRAPPER_URL: https://u7tpgfw3bh.execute-api.eu-west-2.amazonaws.com/dev
+
```
Comment (idrop) [inline template.yaml:173]: can we remove this newline to join the event to the Function pls?

File: template.yaml [hunk 9]
```diff
@@ -127,21 +140,67 @@ Resources:
...
+            TableName: !Ref KBVSessionTable
+        - DynamoDBWritePolicy:
+            TableName: !Ref KBVSessionTable
+      Runtime: java11
+      MemorySize: 512
+      Tracing: Active
+      Environment:
+        Variables:
+          IS_LOCAL: false
+          KBV_SESSION_TABLE_NAME: !Ref KBVSessionTable
```
Comment (idrop) [inline template.yaml:169]: these env vars should probably end up being SSM parameters but ok for now

File: lambdas/question/src/test/java/uk.gov.di.ipv.cri.kbv.api.handler/QuestionHandlerTest.java (hunk 7 above)
Comment (amanm-alphagov) [reply lambdas/question/src/test/java/uk.gov.di.ipv.cri.kbv.api.handler/QuestionHandlerTest.java:95]: Addressed a few from the Major and critical code smells highlighted by SonarCloud

File: lambdas/question/src/main/java/uk.gov.di.ipv.cri.kbv.api.handler/QuestionHandler.java [hunk 10]
```diff
@@ -43,83 +46,83 @@ public QuestionHandler() {
...
         String responseBody = "{}";
         Map<String, String> responseHeaders = Map.of("Content-Type", "application/json");
         String json;
         int statusCode;
 
         try {
             String sessionId = input.getHeaders().get(HEADER_SESSION_ID);
-            KBVSessionItem kbvSessionItem = storageService.getSessionId(sessionId);
+            KBVSessionItem kbvSessionItem =
+                    storageService.getSessionId(sessionId).orElseThrow(NullPointerException::new);
```
Comment (idrop) [inline lambdas/question/src/main/java/uk.gov.di.ipv.cri.kbv.api.handler/QuestionHandler.java:73]: this should probably result in a 400 error and not the server error we're giving it below

File: template.yaml (hunk 8 above)
Comment (amanm-alphagov) [reply template.yaml:173]: Done.

File: lambdas/session/src/main/java/uk/gov/di/ipv/cri/kbv/api/handler/QuestionAnswerHandler.java [hunk 11]
```diff
@@ -115,7 +112,7 @@ private QuestionsResponse submitAnswers(QuestionState questionState)
         HttpResponse<String> res = httpClient.send(httpReq, HttpResponse.BodyHandlers.ofString());
 
         String body = res.body();
-        System.out.println("😁 got rtq response: " + body);
+        System.out.println("got rtq response: " + body);
```
Comment (idrop) [inline lambdas/session/src/main/java/uk/gov/di/ipv/cri/kbv/api/handler/QuestionAnswerHandler.java:115]: @amanm-alphagov one last remaining sout. Can we remove pls?

**Comment by jkunle** [comment]
Test Coverage is low on this repo aim to raise this on the next task, more classes should be covered in future and endeavor to try to resolve the remaining code-smell on the next task

**Review by jkunle — APPROVED** [review]

**Review by idrop — APPROVED** [review]

**Pull request merged**